pallets/liquid-staking/src/lib.rs

```diff
@@ -241,70 +256,15 @@ pub mod pallet {
     impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
         /// Try to pay off over the `UnstakeQueue` while blockchain is on idle.
-        ///
-        /// It breaks when:
-        ///     - Pallet's balance is insufficiant.
-        ///     - Queue is empty.
-        ///     - `remaining_weight` is less than one pop_queue needed.
-        fn on_idle(_n: BlockNumberFor<T>, mut remaining_weight: Weight) -> Weight {
-            // on_idle shouldn't run out of all remaining_weight normally
+        fn on_idle(_n: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
             let base_weight = <T as Config>::WeightInfo::on_idle();
-            let staking_currency = Self::staking_currency();
-
-            // return if staking_currency haven't been set.
-            if staking_currency.is_err() {
+            if remaining_weight < base_weight {
                 return remaining_weight;
             }
-
-            let staking_currency = staking_currency.expect("It must be ok; qed");
-
-            loop {
-                // check weight is enough
-                if remaining_weight < base_weight {
-                    break;
-                }
-
-                if Self::unstake_queue().is_empty() {
-                    break;
-                }
-
-                // get the front of the queue.
-                let (who, amount) = &Self::unstake_queue()[0];
-                let account_id = Self::account_id();
-
-                // InsurancePool should not be embazzled.
-                let free_balance =
-                    T::Assets::reducible_balance(staking_currency, &account_id, false)
-                        .saturating_sub(Self::total_reserves());
-
-                log::trace!(
-                    target: "liquidstaking::on_idle",
-                    "account: {:?}, unstake_amount: {:?}, remaining_weight: {:?}, pallet_free_balance: {:?}",
-                    who,
-                    amount,
-                    remaining_weight,
-                    free_balance,
-                );
-                if free_balance < *amount {
-                    return remaining_weight;
-                }
-
-                if let Err(err) =
-                    T::Assets::transfer(staking_currency, &account_id, who, *amount, false)
-                {
-                    log::error!(target: "liquidstaking::on_idle", "Transfer failed {:?}", err);
-                    // break if we cannot afford this
-                    break;
-                }
-
-                // substract weight of this action if succeed.
-                remaining_weight -= base_weight;
-
-                // remove unstake request from queue
-                Self::pop_unstake_task()
-            }
-
-            remaining_weight
+            with_transaction(|| match Self::do_pop_front() {
+                Ok(_) => TransactionOutcome::Commit(remaining_weight - base_weight),
+                Err(_) => TransactionOutcome::Rollback(0),
+            })
```

Comment:
Seems like now we only process one account each time, due to the `unstake queue cap` and `minunstake amount`, maybe this would be the throughput for the system.